### ISSUE TITLE
filter_kubernetes: API Server calls are ignored sometimes

### ISSUE THREAD

**Issue by alitoufighi** [fluent/fluent-bit#2249]
## Bug Report

**Describe the bug**
As I see in how Kubernetes Filter works, it fetches Pod metadata (Pod ID, labels, etc.) from apiserver once for a <Pod, Namespace> pair, and caches it for later usages. But for some of Pods, it seems to be ignored and no metadata and no labels are added to the log object. Also no logs are present to mention what the problem is (even in debug level).

It's good to mention that I also experience casual apiserver connection errors:
```
[error] [filter:kubernetes:kubernetes.0] upstream connection error
```
But its frequency is much less than incomplete logs I see in output. However, I suspect that it may have some effects on the behavior.

**To Reproduce**
I don't know the cause of the error, but **all** logs for some of my Pods lack metadata:
```
[0] kube.var.log.containers.XXXX-856bf98d7c-pfhp5_production_XXXX-b0222ff276f50e2b952ba69215aa0333e056d2c189a6ae0f0a0ff909bd77b428.log: 
[1591913404.804862430, {"log"=>"{"XXX"}", "time"=>"2020-06-11T22:10:04.80486243Z", "pod_name"=>"XXXX-856bf98d7c-pfhp5", "namespace_name"=>"production"}]
``` 
**NOTE:** While writing this issue, I found that all of these ignored Pods are from some specific Deployments, but not all of their Pods are experiencing this.

**Expected behavior**
Include metadata in the log object, such that the output would be like this (host and container_image are added for example):
```
[0] kube.var.log.containers.XXXX-856bf98d7c-pfhp5_production_XXXX-b0222ff276f50e2b952ba69215aa0333e056d2c189a6ae0f0a0ff909bd77b428.log: 
[1591913404.804862430, {"log"=>"{"XXX"}", "time"=>"2020-06-11T22:10:04.80486243Z", "pod_name"=>"XXXX-856bf98d7c-pfhp5", "namespace_name"=>"production",
"host"=>"XXXX", "container_image"=>"XXXX"}]
``` 

**Configuration File**
```
[SERVICE]
    Flush         1
    Log_Level     info
    Daemon        off
    Parsers_File  parsers.conf

[INPUT]
    Name               tail
    Tag                kube.*
    Path               /var/log/containers/*.log
    Parser             docker
    DB                 /var/log/flb_kube.db
    Mem_Buf_Limit      5MB
    Skip_Long_Lines    Off
    Docker_Mode        On
    Refresh_Interval   10

[FILTER]
    Name         record_modifier
    Match        *
    Remove_key   pod-template-hash

[FILTER]
    Name         record_modifier
    Match        *
    Remove_key   container_name

[FILTER]
    Name         record_modifier
    Match        *
    Remove_key   stream

[FILTER]
    Name         record_modifier
    Match        *
    Remove_key   docker_id

[FILTER]
    Name         record_modifier
    Match        *
    Remove_key   container_hash

[FILTER]
    Name         record_modifier
    Match        *
    Remove_key   pod_id

[PARSER]
    Name        docker
    Format      json
    Time_Key    time
    Time_Format %Y-%m-%dT%H:%M:%S.%L
    Time_Keep   On

[OUTPUT]
    Name    stdout
    Match   *
```

**Your Environment**
* Version used: `fluent/fluent-bit:1.4.1`
* Configuration: As above
* Environment name and version: Kubernetes 1.18 on Docker 19.3.8
* Server type and version: VMWare Virtual Machines
* Operating System and version: Ubuntu 18.04
* Filters and plugins: As above

**Comment by fujimotos** [comment]
> It's good to mention that I also experience casual apiserver connection errors:

@alitoufighi Fluent Bit checks the connectivity during the
initialization. If it can't connect to the API server, it assumes
that kubernetes filter is unusable.

It depends on the exact reason why it failed to connect on your env,
but I am guessing that the fix ded63b77 can resolve your issue too.

This fix is available since v1.4.5. Can you test that version and
see if it resolves your issue?

**Comment by alitoufighi** [comment]
Thanks @fujimotos, I upgraded to 1.4.6.

This reduced the happenstances a lot, but not completely. Before upgrading I was facing this issue almost 10 times per second, but now it's about 2 times per minute, and these 2 are happening in one moment.

I don't know if it's a problem with my DNS lookups or connectivity issues in my cluster or it's Fluent Bit that has issues, but running `nslookup kubernetes.default.svc` once per second in one of my Pods for about 12 hours showed no errors.

By investigating I found that after upgrade, the behavior has been changed. Before upgrading I had `pod_name` and `namespace_name` fields present (at least in those logs that I saw having problems), but now in these remaining instances, none of them are present. (Why?!)
Also I found that these bulk failures are related to Pods that are newly created and **maybe API Server has delays in responding the metadata** or something for newly created Pods?

For example, these three lines are logged right after each other:
```
[2020/06/17 16:20:29] [ info] inotify_fs_add(): inode=10391957 watch_fd=1029 name=/var/log/containers/XXX-6d849f97bc-lmbls_production_XXX-aec2cfddf01c1fd013fe38bf874364a8a2d2a0092c45de44ffdb3e5f51e3434c.log
[2020/06/17 16:20:30] [error] [filter:kubernetes:kubernetes.0] upstream connection error
[0] kube.var.log.containers.XXX-6d849f97bc-lmbls_production_XXX-aec2cfddf01c1fd013fe38bf874364a8a2d2a0092c45de44ffdb3e5f51e3434c.log: [1592394630.459574275, {"log"=>"foobar"}]  # Lacks K8s Filter fields 
```

Do you have any ideas on this?
